File: medium/173. Binary Search Tree Iterator/173. Binary Search Tree Iterator.ts
/**
 * Definition for a binary tree node.
 */

class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}

class BSTIterator {
    stack: TreeNode[]

    constructor(root: TreeNode | null) {
        this.stack = []
        this.mostLeftInOrder(root)
    }

    mostLeftInOrder(tree: TreeNode | null) {
        if (!tree) return

        while (tree) {
            this.stack.push(tree)
            tree = tree.left
        }
    }

    next(): number {
        const topOfTheStack = this.stack.pop()

        if (!topOfTheStack)
            return -1

        if (topOfTheStack?.right)
            this.mostLeftInOrder(topOfTheStack.right)
        return topOfTheStack.val
    }

    hasNext(): boolean {
        return this.stack.length > 0
    }
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * var obj = new BSTIterator(root)
 * var param_1 = obj.next()
 * var param_2 = obj.hasNext()
 */

File: medium/173. Binary Search Tree Iterator/173. Binary Search Tree Iterator.py
from typing import List, Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class BSTIterator:

    def __init__(self, root: Optional[TreeNode]):
        # def transform_tree(tree: Optional[TreeNode]) -> List[int]:
        #     if not tree:
        #         return []

        #     return transform_tree(tree.left) + [tree.val] + transform_tree(tree.right)

        # self.tree_arr = transform_tree(root)
        # self.curr_index = 0
        
        self.stack: List[TreeNode] = []
        self.left_most_in_order(root)

    def left_most_in_order(self, tree: Optional[TreeNode]):
        while tree:
            self.stack.append(tree)
            tree = tree.left
    
    def next(self) -> int:
        top_of_stack = self.stack.pop()
        if top_of_stack.right:
            self.left_most_in_order(top_of_stack.right)
        return top_of_stack.val

    def hasNext(self) -> bool:
        return len(self.stack) > 0


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()
